File: ios/LowerBackStretching/Views/Components/DifficultyDots.swift
import SwiftUI

/// Renders 1–3 filled dots for a difficulty string. Unknown values render
/// one dot. Accessibility label still reads the difficulty name so
/// VoiceOver doesn't lose the information.
struct DifficultyDots: View {
    let difficulty: String

    private var filled: Int {
        switch difficulty.lowercased() {
        case "easy":   return 1
        case "medium": return 2
        case "hard":   return 3
        default:       return 1
        }
    }

    var body: some View {
        HStack(spacing: 3) {
            ForEach(0..<3, id: \.self) { i in
                Circle()
                    .fill(i < filled ? Color.accentColor : Color.primary.opacity(0.2))
                    .overlay(
                        Circle().stroke(
                            i < filled ? Color.clear : Color.primary.opacity(0.35),
                            lineWidth: 1
                        )
                    )
                    .frame(width: 8, height: 8)
            }
        }
        .accessibilityElement()
        .accessibilityLabel(Text(difficulty.capitalized))
    }
}

File: android/app/src/main/java/com/lowerbackstretching/ui/components/DifficultyDots.kt
package com.lowerbackstretching.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

/**
 * Renders 1–3 filled dots for a difficulty string. Unknown values render
 * one dot. Accessibility label still reads the difficulty name so screen
 * readers don't lose the information.
 */
@Composable
fun DifficultyDots(difficulty: String, modifier: Modifier = Modifier) {
    val filled = filledCount(difficulty)
    val label = difficulty.replaceFirstChar(Char::titlecase)
    Row(
        horizontalArrangement = Arrangement.spacedBy(3.dp),
        modifier = modifier.semantics { contentDescription = label },
    ) {
        repeat(3) { i ->
            val isFilled = i < filled
            val color = if (isFilled) MaterialTheme.colorScheme.primary
                        else MaterialTheme.colorScheme.onSurface.copy(alpha = 0.2f)
            Row(
                modifier = Modifier
                    .size(8.dp)
                    .clip(CircleShape)
                    .background(color)
                    .then(
                        if (!isFilled) Modifier.border(
                            width = 1.dp,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.35f),
                            shape = CircleShape,
                        ) else Modifier
                    ),
                content = {},
            )
        }
    }
}

internal fun filledCount(difficulty: String): Int = when (difficulty.lowercase()) {
    "easy" -> 1
    "medium" -> 2
    "hard" -> 3
    else -> 1
}
